### PR TITLE
Possible improvements for the concurrent queue

### PR DESCRIPTION
**Pull request by Scooletz**
This is a highly experimental PR. A few optimizations could be considered like:

- using `ref` locals for slots. Once obtained they could be used
- once I understand how to do partials for CoreCLR (I saw https://github.com/dotnet/corefx/tree/master/src/Common/src/CoreLib#conventions ), possibly consider using Unsafe for accessing
- pseudo-randomize access to slots (use bitwise operators mix indexes and lower a possibility of cache trashing)

This needs benchmarks for at least two scenarios MPMC SPSC as they might be impaired by the changes.

### CODE REVIEW

**Comment by danmosemsft** [comment]
Thanks! There should be benchmarks (and good docs on how to run them) in the dotnet/performance repo. We are always interested in perf improvements, supported by data.

File: src/Common/src/CoreLib/System/Collections/Concurrent/ConcurrentQueueSegment.cs [hunk 1]
```diff
@@ -138,9 +138,10 @@ internal static int RoundUpToPowerOf2(int i)
                 // Get the head at which to try to dequeue.
                 int currentHead = Volatile.Read(ref _headAndTail.Head);
                 int slotsIndex = currentHead & _slotsMask;
+                ref Slot slot = ref slots[slotsIndex];
```
Comment (stephentoub) [inline src/Common/src/CoreLib/System/Collections/Concurrent/ConcurrentQueueSegment.cs]: FWIW, I'd previously tried this use of ref and didn't see any measurable benefits, but you're off course welcome to try again; things may have changed in the interim.

However, ConcurrentQueue itself is built in the coreclr repo, so any changes to this Segment here won't actually be observable as part of CQ benchmarking. It'd be better to experiment in the coreclr repo. It's also not really worth a PR until you have some measurements to justify the changes.

Thanks!

Comment (Scooletz) [reply src/Common/src/CoreLib/System/Collections/Concurrent/ConcurrentQueueSegment.cs]: Closing it then. Will raise when a potential gain is observed. Thank you @stephentoub

**Comment by Scooletz** [comment]
Closed for now. If a gain is observed, will be raised.

**Pull request closed without merging**